accounts-db/src/accounts_db/geyser_plugin_utils.rs

```diff
@@ -89,40 +89,65 @@ impl AccountsDb {
     ) {
         let storage_entry = self.storage.get_slot_storage_entry(slot).unwrap();
 
-        let mut accounts_to_stream: HashMap<Pubkey, StoredAccountMeta> = HashMap::default();
+        let mut accounts_duplicate: HashMap<Pubkey, usize> = HashMap::default();
         let mut measure_filter = Measure::start("accountsdb-plugin-filtering-accounts");
-        let accounts = storage_entry.accounts.account_iter();
         let mut account_len = 0;
-        accounts.for_each(|account| {
+        let mut pubkeys = HashSet::new();
+
+        // populate `accounts_duplicate` for any pubkeys that are in this storage twice.
+        // Storages cannot return `StoredAccountMeta<'_>` for more than 1 account at a time, so we have to do 2 passes to make sure
+        // we don't have duplicate pubkeys.
+        let mut i = 0;
+        storage_entry.accounts.scan_pubkeys(|pubkey| {
+            i += 1; // pre-increment to most easily match early returns in next loop
+            if !pubkeys.insert(*pubkey) {
+                accounts_duplicate.insert(*pubkey, i); // remember the highest index entry in this slot
+            }
+        });
+
+        // now, actually notify geyser
+        let mut i = 0;
+        storage_entry.accounts.scan_accounts(|account| {
+            i += 1;
             account_len += 1;
             if notified_accounts.contains(account.pubkey()) {
                 notify_stats.skipped_accounts += 1;
                 return;
             }
+            if let Some(highest_i) = accounts_duplicate.get(account.pubkey()) {
+                if highest_i != &i {
+                    // this pubkey is in this storage twice and the current instance is not the last one, so we skip it.
+                    // We only send unique accounts in this slot to `notify_filtered_accounts`
+                    return;
+                }
```

Comment:
i wondered this, too. in the previous impl, we incremented `account_len` for each account in `accounts.for_each(|account| {`. This would have included duplicates already. The impl in this pr should be exactly the same behavior as the previous impl.